Design for per-package documentation system (#306)
ros2/design: articles/per_package_documentation.md

**2.7 Changes to documentation pushed in a package repository should be immediately deployed to the documentation site**

**2.8 The system should support bi-directional cross-referencing (“backwards” to dependents)**

## Design

The design covers the solution for satisfying the primary requirements.
Some secondary requirements are also satisfied inherently through the primary design decisions.

In summary, the design for the system is:

- Utilize [Sphinx][1], [Doxygen][2] and [Breathe][3] tools to satisfy both C++ and Python API documentation (1.4), cross-referencing across packages (1.9) and consistent use of RST (1.5)
- Develop a new doc tool integrated into the build system (1.7) for maintainers to run on their packages (1.3) that encompasses Sphinx, Doxygen and Breathe, and builds the docs locally (2.3).
- The output of the tool is uploaded to docs.ros.org (1.1) by the build farm, to the correct URL address (1.2, 1.10)
- A package’s inclusion in a rosdistro (as per normal release process*) will version its documentation to that distribution (1.6) and list the package under that distribution’s listing of packages on docs.ros.org (1.8)

>     - With the inclusion of a doc block in distribution.yaml

Reviewer: ```suggestion
    - Via the inclusion of a doc block in a distribution.yaml in https://github.com/ros/rosdistro
```

#### Buildfarm process

To opt in to having documentation built on docs.ros.org, maintainers will have to add a `doc` block to their `distribution.yaml` file that indicates the rosdistro branch and a path to the repository.
Opting-in this way will cause the tool to automatically generate an index landing page for the package.

The buildfarm will invoke the same tool used in the user process, run on ***n*** packages, to generate the documentation for `docs.ros.org`.
It will require some additional settings, for example the URL where content should generate into.

A webhook between GitHub and the buildfarm will indicate when there is a new commit to a repository (for the branches being tracked for documentation) so the documentation can be automatically regenerated and immediately deployed.

Some files will be auto-generated during this process (changelogs, tag files, etc.).
The auto-generated content won’t be saved to the upstream repositories, but it will be built to `docs.ros.org`.

There is a possibility of collision occurring between maintainer created files and generated files.

> To prevent this to the best of the system's ability, the system can place auto-generated content in a restricted directory (see [URL Structure - /generated](#generated)).

Reviewer: ```suggestion
To prevent this to the best of the system's ability, the system will place auto-generated content in a restricted directory (see [URL Structure - /generated](#generated)).
```

### URL structure

![The URL structure of the new documentation system](per_package_documentation/url-structure.png)

#### `index` pages

The index pages indicate the default page that is rendered when the directory level immediately before is visited without any specific `.html` page following.

For example, visiting docs.ros.org/en/foxy would render `index1`, docs.ros.org/en/foxy/p would render `index2`, and docs.ros.org/en/foxy/p/<package_name> would render `index3`.

`index1` is the general ROS 2 documentation overview page, what is currently seen when visiting index.ros.org/doc/ros2/

`index2` will be the package listing page for each ROS 2 distribution.

`index3` will be each specific package’s homepage.
This will be generated by the tool, but a user-generated index page should also be able to be merged or augmented with the auto-generated page to prevent collision.

#### Versioning at the `<package_name>` level

> To support the goal of being able to support multiple versions of packages documented within  the context of one distro (2.6) we plan to reserve URL space and have a plan to migrate forward but not to implement it at first.

Reviewer: ```suggestion
To support the goal of being able to support multiple versions of packages documented within the context of one distro (2.6) we plan to reserve URL space and have a plan to migrate forward but not to implement it at first.
```
